src/lib.rs

```diff
@@ -1,772 +1,331 @@
 //! Tarjan's zip tree implementation in Rust.
 //!
-//! Zip tree is a treap with different insertion and deletion algorithms.
-//! It organizes node ranks like skip list, but takes less space than skip list.
-//! Insertion and deletion are done by _zip_ and _unzip_ operations instead of
-//! a series of tree rotations. You can see [Tarjans's paper](https://arxiv.org/abs/1806.06726)
+//! Zip tree can be seen as a variant of treap with different node rank distributions
+//! and insertion and deletion algorithms. It is isomorphic to skip list, but
+//! takes less space as each node only needs to store its left and right subtrees
+//! and O(log log n) bit for its rank. Insertion and deletion are done by _zip_
+//! and _unzip_ operations instead of a series of tree rotations. You can see
+//! [1](https://arxiv.org/abs/1806.06726) and [2](https://arxiv.org/abs/2307.07660)
 //! for more details.
 
-extern crate rand;
-
-use std::marker::PhantomData;
-use std::ops::{Index, IndexMut};
-use std::borrow::Borrow;
-use std::mem::replace;
+use rand::distributions::{Distribution, Standard};
+use rand::rngs::SmallRng;
+use rand::{Rng, SeedableRng};
+use std::borrow::{Borrow, BorrowMut};
 use std::cmp::Ordering;
-use std::collections::{LinkedList, HashMap};
+use std::collections::VecDeque;
 use std::fmt::Debug;
-use std::ptr::null_mut;
-use rand::prelude::*;
+use std::ops::{Bound, RangeBounds};
 
-/// The struct is created by [ZipTree::into_iter()](struct.ZipTree.html#impl-IntoIterator).
-#[derive(Clone)]
-pub struct IntoIter<K, V> {
-    iter: std::collections::linked_list::IntoIter<(K, V)>,
-}
+#[derive(PartialEq, Eq, PartialOrd, Ord, Debug, Clone)]
+struct Rank(u16);
 
-/// The struct is created by [ZipTree::iter()](struct.ZipTree.html#method.iter).
-#[derive(Clone)]
-pub struct Iter<'a, K: 'a, V: 'a> {
-    tree: &'a ZipTree<K, V>,
-    path: LinkedList<(*mut Node<K, V>, DfsAction)>,
-    dummy: PhantomData<(&'a K, &'a V)>,
+impl Distribution<Rank> for Standard {
+    #[inline]
+    fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> Rank {
+        let r1 = rng.gen::<u64>().leading_zeros() as u16;
+        let r2 = rng.gen::<u16>() & 0x3FFu16;
+        Rank((r1 << 10) + r2)
```

Comment:
Can you please explain the logic here? Everywhere I look in the original paper r1 and r2 are to be stored separately, and I don't understand why that bitwise AND and why that number

Reply:
This stores `r1` in the first 6 bits and `r2` in the last 10 bits. We need to `& 0x3FFu16` to make the first 6 bits of `r2` 0.